add topn-limit-push-down (#3089)
pingcap/docs: topn-limit-push-down.md

## Examples

The following describes TopN pushdown through some examples.

### Example 1: Push down to the Coprocessor in the storage layer

{{< copyable "sql" >}}

```sql
create table t(id int primary key, a int not null);
explain select * from t order by a limit 10;
```

```
+----------------------------+----------+-----------+---------------+--------------------------------+
| id                         | estRows  | task      | access object | operator info                  |
+----------------------------+----------+-----------+---------------+--------------------------------+
| TopN_7                     | 10.00    | root      |               | test.t.a, offset:0, count:10   |
| └─TableReader_15           | 10.00    | root      |               | data:TopN_14                   |
|   └─TopN_14                | 10.00    | cop[tikv] |               | test.t.a, offset:0, count:10   |
|     └─TableFullScan_13     | 10000.00 | cop[tikv] | table:t       | keep order:false, stats:pseudo |
+----------------------------+----------+-----------+---------------+--------------------------------+
4 rows in set (0.00 sec)
```

> In this query, the TopN operator node was pushed down to TiKV for data filtering, and each Coprocessor returns only 10 records to TiDB. After TiDB integrates the data, the final filtering is performed.

Reviewer: ```suggestion
In this query, the TopN operator node is pushed down to TiKV for data filtering, and each Coprocessor returns only 10 records to TiDB. After TiDB aggregates the data, the final filtering is performed.
```

> ### Example 2: TopN can be pushed down into the Join (sorting only depends on the columns in the outer table)

Reviewer: ```suggestion
### Example 2: TopN can be pushed down into Join (sorting only depends on the columns in the outer table)
```

| TopN_12                          | 10.00    | root      |               | test.t.a, offset:0, count:10                    |
| └─HashJoin_17                    | 12.50    | root      |               | left outer join, equal:[eq(test.t.a, test.s.a)] |
|   ├─TopN_18(Build)               | 10.00    | root      |               | test.t.a, offset:0, count:10                    |
|   │ └─TableReader_26             | 10.00    | root      |               | data:TopN_25                                    |
|   │   └─TopN_25                  | 10.00    | cop[tikv] |               | test.t.a, offset:0, count:10                    |
|   │     └─TableFullScan_24       | 10000.00 | cop[tikv] | table:t       | keep order:false, stats:pseudo                  |
|   └─TableReader_30(Probe)        | 10000.00 | root      |               | data:TableFullScan_29                           |
|     └─TableFullScan_29           | 10000.00 | cop[tikv] | table:s       | keep order:false, stats:pseudo                  |
+----------------------------------+----------+-----------+---------------+-------------------------------------------------+
8 rows in set (0.01 sec)
```

> In this query, the sorting of the TopN operator only depends on the columns in the outer table t, so TopN can be pushed down before Join, to reduce the computational of Join. Besides, TiDB also pushed TopN down to the storage layer.

Reviewer: ```suggestion
In this query, the collation of the TopN operator only depends on the columns in the outer table `t`, so a calculation can be performed before pushing down TopN to Join, to reduce the calculation cost of the Join operation. Besides, TiDB also pushes TopN down to the storage layer.
```
We usually translate `排序规则` as `collation`~

| id                            | estRows  | task      | access object | operator info                              |
+-------------------------------+----------+-----------+---------------+--------------------------------------------+
| TopN_12                       | 10.00    | root      |               | test.t.id, offset:0, count:10              |
| └─HashJoin_16                 | 12500.00 | root      |               | inner join, equal:[eq(test.t.a, test.s.a)] |
|   ├─TableReader_21(Build)     | 10000.00 | root      |               | data:TableFullScan_20                      |
|   │ └─TableFullScan_20        | 10000.00 | cop[tikv] | table:s       | keep order:false, stats:pseudo             |
|   └─TableReader_19(Probe)     | 10000.00 | root      |               | data:TableFullScan_18                      |
|     └─TableFullScan_18        | 10000.00 | cop[tikv] | table:t       | keep order:false, stats:pseudo             |
+-------------------------------+----------+-----------+---------------+--------------------------------------------+
6 rows in set (0.00 sec)
```

> TopN can't be pushed down into the Inner Join. Taking the query above as an example, if you get 100 records after Join, then do TopN can leave 10 records. If the remaining 10 records are filtered before TopN, there may be 5 left after the Join is done. It will make results differences.

Reviewer: ```suggestion
TopN cannot be pushed down before `Inner Join`. Taking the query above as an example, if you get 100 records after Join, then you can have 10 records left after TopN. However, if TopN is performed first to get 10 records, only 5 records are left after Join. The pushdown might result in different results. 
```
Please check if this is technically correct.

> That is the reason why opN can't be pushed down into the Outer Join for the Inner table. TopN can not be pushed down when sorting by multiple columns on multiple tables. such as `t.a+s.a`.Only if the TopN exclusively depends on sorting outer table columns, it can be pushed down. 

Reviewer: ```suggestion
Similarly, TopN can neither be pushed down into the Outer Join for the inner table, nor can it be pushed down when its collation is related to columns on multiple tables, such as `t.a+s.a`. Only when the collation of TopN exclusively depends on columns on the outer table, can it be pushed down. 
```

> ### Example 4: Conversion of TopN to Limit

Reviewer: ```suggestion
### Example 4: Convert TopN to Limit
```

| TopN_12                          | 10.00    | root      |               | test.t.id, offset:0, count:10                   |
| └─HashJoin_17                    | 12.50    | root      |               | left outer join, equal:[eq(test.t.a, test.s.a)] |
|   ├─Limit_21(Build)              | 10.00    | root      |               | offset:0, count:10                              |
|   │ └─TableReader_31             | 10.00    | root      |               | data:Limit_30                                   |
|   │   └─Limit_30                 | 10.00    | cop[tikv] |               | offset:0, count:10                              |
|   │     └─TableFullScan_29       | 10.00    | cop[tikv] | table:t       | keep order:true, stats:pseudo                   |
|   └─TableReader_35(Probe)        | 10000.00 | root      |               | data:TableFullScan_34                           |
|     └─TableFullScan_34           | 10000.00 | cop[tikv] | table:s       | keep order:false, stats:pseudo                  |
+----------------------------------+----------+-----------+---------------+-------------------------------------------------+
8 rows in set (0.00 sec)

```

> In the above query, TopN is first pushed to the outer table t. Because TopN needs to sort by `t.id` which is the primary key,  it can be directly readout in order  (`keep order: true`) without extra sorting in TopN, and Simplified to Limit.

Reviewer: ```suggestion
In the query above, TopN is first pushed to the outer table `t`. TopN needs to sort by `t.id`, which is the primary key and can be directly read in order  (`keep order: true`) without extra sorting in TopN. Therefore, TopN is simplied as Limit.
```

---
title: TopN and Limit Operator Pushdown
summary: Learn the implementation of TopN and Limit operator pushdown.

> category: performance

Reviewer 2: ```suggestion
```

Reviewer 3: > > It's better to replace `collation` with `ordering rule` to avoid misunderstandings with the collation of sql.
> 
> Is there any ref on this one? Like, how would you choose between `ordering rule` and `sorting rule`?

No, `collation` generally refers to the collation of the character set. You can choose from other options that you think are more appropriate.
